scapy/contrib/cansocket_python_can.py

```diff
@@ -334,9 +310,19 @@ def select(sockets, remain=conf.recv_poll_rate):
         :returns: an array of sockets that were selected and
             the function to be called next to get the packets (i.g. recv)
         """
+        ready_sockets = \
+            [s for s in sockets if isinstance(s, PythonCANSocket) and
+             len(s.can_iface.rx_queue)]
+        # checking the queue length without locking might sound
+        # dangerous, but for the purpose of this select, if another
+        # thread is reading the same socket, then even proper locking
+        # wouldn't help
+        if not len(ready_sockets):
+            # yield this thread to avoid starvation
+            time.sleep(0)
```

Comment:
I've never seen this. What does it do?

Reply:
Well, we observed the following behavior, when using bridge and sniff. Here is our testcase:

```python


sock0 = CANSocket(bustype='socketcan', channel='vcan0')
sock1 = CANSocket(bustype='socketcan', channel='vcan1')

bridgeStarted = threading.Event()
def bridge():
    global bridgeStarted
    bSock0 = CANSocket(
        bustype='socketcan', channel='vcan0',
                                bitrate=250000)
    bSock1 = CANSocket(
        bustype='socketcan', channel='vcan1',
                                bitrate=250000)
    def pnr(pkt):
        return pkt
    bSock0.timeout = 0.01
    bSock1.timeout = 0.01
    bridge_and_sniff(if1=bSock0, if2=bSock1, xfrm12=pnr, xfrm21=pnr, timeout=0.5, started_callback=bridgeStarted.set, count=6)
    bSock0.close()
    bSock1.close()

threadBridge = threading.Thread(target=bridge)
threadBridge.start()
bridgeStarted.wait(timeout=1)

sock0.send(CAN(flags='extended', identifier=0x10010000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))
sock0.send(CAN(flags='extended', identifier=0x10020000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))
sock0.send(CAN(flags='extended', identifier=0x10000000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))
sock0.send(CAN(flags='extended', identifier=0x10030000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))
sock0.send(CAN(flags='extended', identifier=0x10040000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))
sock0.send(CAN(flags='extended', identifier=0x10000000, length=8, data=b'\x01\x02\x03\x04\x05\x06\x07\x08'))

packetsVCan1 = sock1.sniff(timeout=0.5, count=6)
assert len(packetsVCan1) == 6

sock1.close()
sock0.close()

threadBridge.join(timeout=3)
assert not threadBridge.is_alive()

```
Basically, a background thread runs a bridge and the foreground thread executes some send commands on one off the sockets, connected by the bridge.

We observed the error, that the foreground thread didn't received any computing time, since the bridge thread is constantly trying to receive messages. Inside the bridge, the send_recv function calls select in a while loop. Without the `time.sleep(0)` the background thread will not give any computing resources away, which results, in an testcase error. The bridge will basically timeout, then all messages are sent. 

Feel free to test this behavior yourself. 

Once the time.sleep(0) is removed, this unit-test will randomly fail:

```
sudo ./test/run_tests -3 -F -t test/contrib/cansocket_python_can.uts
```